### DISCUSSION TITLE
How to calculate the area of Alpha _ shape _ 2 or Alpha _ shape _ 3?

### DISCUSSION THREAD

**Discussion by Kasi** [CGAL/cgal#7569] [Q&A]
How to calculate the area of Alpha _ shape _ 2 or Alpha _ shape _ 3?

**Comment by stla** [comment]
What do you mean by the area for the 3D case? The area of the surface? You can make a mesh with the alpha shape and there's a function in CGAL to get the area of the surface of a 3D mesh.

**Reply by Kasi** [reply]
Use this function CGAL:: Polygon_mesh_processing:: volume() but how to make a 3D mesh with alpha shape?

Thank you very much for your reply!

**Reply by stla** [reply]
I use **Rcpp** (a C++ library for R) so you will have to adapt my code. Once you have the alpha shape in `as` you get its vertices as follows and the face indices are `3i, 3i+1, 3i+2`:

```cpp
  std::list<Facet> facets;
  as.get_alpha_shape_facets(std::back_inserter(facets),
                            Fixed_alpha_shape_3::REGULAR);
  as.get_alpha_shape_facets(std::back_inserter(facets),
                            Fixed_alpha_shape_3::SINGULAR);
  const int nfacets = facets.size();
  // output - facet indices are 3i, 3i+1, 3i+2
  Rcpp::NumericMatrix Vertices(3, 3 * nfacets);
  std::list<Facet>::iterator it_facet;
  int i = 0;
  for(it_facet = facets.begin(); it_facet != facets.end(); it_facet++) {
    Facet facet = *it_facet;
    // to have a consistent orientation, always consider an exterior cell
    if(as.classify(facet.first) != Fixed_alpha_shape_3::EXTERIOR) {
      facet = as.mirror_facet(facet);
    }
    int indices[3] = {
      (facet.second + 1) % 4,
      (facet.second + 2) % 4,
      (facet.second + 3) % 4,
    };
    // needed to get a consistent orientation
    if(facet.second % 2 == 0 ){
      std::swap(indices[0], indices[1]);
    }
    const Point3 v1 = facet.first->vertex(indices[0])->point();
    const Point3 v2 = facet.first->vertex(indices[1])->point();
    const Point3 v3 = facet.first->vertex(indices[2])->point();
    const Rcpp::NumericVector V1 = {v1.x(), v1.y(), v1.z()};
    const Rcpp::NumericVector V2 = {v2.x(), v2.y(), v2.z()};
    const Rcpp::NumericVector V3 = {v3.x(), v3.y(), v3.z()};
    Vertices(Rcpp::_, 3*i)   = V1;
    Vertices(Rcpp::_, 3*i+1) = V2;
    Vertices(Rcpp::_, 3*i+2) = V3;
    i++;
  }
```

**Reply by Kasi** [reply]
thanks，But,
[https://github.com/stla/AlphaHull3D/blob/main/src/alphahull3d.cpp](url)
Why is the alpha shape calculating the same volume and the hull volume?
`Polyhedron p;
CGAL::convex_hull_3(points.begin(), points.end(), p);
CGAL::Polygon_mesh_processing::volume(p);`

**Reply by stla** [reply]
I don't understand your question. The alpha shape is not convex in general, so it makes sense to calculate the volume of its convex hull (but I don't remember I did that).

**Reply by Kasi** [reply]
Yes, it's very strange.alpha shape Are concave, hull are convex but they have the same result value!

**Reply by stla** [reply]
Not possible. The alpha shape is smaller than the convex hull if it is not equal to the convex hull.

**Reply by Kasi** [reply]
Yes.
I solved the problem.
`as.set_alpha(1.0)`
why?

**Reply by stla** [reply]
Your questions are not understandable. Please make an effort to be clearer. I don't know what you're talking about.

**Reply by Kasi** [reply]
Sorry, English is not my mother tongue. I am using the translation software.

**Reply by Kasi** [reply]
I set the alpha value solved the problem that the volume of alpha shape equals the volume of convex hull.
as.set_alpha(1.0)

Why do I need to set the alpha value?

**Reply by sloriot** [reply]
I don't know what is your shape but for any shape, there exists a value of alpha such that the alpha shape is actually the convex hull. So it might be that 1 is larger or equal to that value for your shape.

**Reply by stla** [reply]
> Why do I need to set the alpha value?

The alpha shape depends on a parameter alpha. Take a look at the documentation. When the alpha value is big enough, the alpha shape is the convex hull. You don't visualize your meshes?
